rfcs/react-components/convergence/slot-render-as.md

```diff
@@ -0,0 +1,155 @@
+  <AccordionHeader button={{ children: (Component, props) => <Component {...props} /> }} />
+
+  {/* New behavior, use 'as' instead, and `props` includes the slot's default `children`. */}
+  <AccordionHeader button={{ as: (Component, props) => <Component {...props} /> }} />
+</>
+```
+
+You can still pass in a string to the `as` prop (e.g. `as="a"` to render an `<a>` tag), and it would work as it does today:
+
+- If the slot's type is an intrinsic element like `'button'`, it replaces that element with the given one.
+- If the slot's type is a component like `Button`, it forwards the `as` prop to the component.
+
+> 🤔 What if you want to provide a render function _and_ override the component's `as` prop? Write a render function that adds your own `as` and spreads props:
+>
+> ```jsx
+> <Checkbox label={{ children: 'Label text', as: (_ignored, props) => <MyLabel as="span" {...props} /> }} />
+> ```
+
+### Pros
+
+1. It allows us to pass the slot's default `children` to the render function, because it doesn't require overriding the default children.
+
+   ```jsx
+   <AccordionHeader
+     button={{
+       as: (Component, props) => (
+         <Component {...props}>
+           <div>extra child</div>
+           {...props.children}
+         </Component>
+       ),
+     }}
+   />
+   ```
+
+2. It allows the component to use children as a function if it needs to, without interfering with the ability for the user to write a slot render function for the root slot:
+
+   ```jsx
+   <Field as={(_ignored, props) => <MyCustomRoot foo="bar" {...props} />}>
+     {inputProps => (
+       <div>
+         <input {...inputProps} />
+       </div>
+     )}
+   </Field>
+   ```
+
+### Cons
+
+1. It overloads the `as` prop, which may run into issues in implementation, if there are any other uses of `as` as a function.
+   - One alternative would be to add a `render` prop instead, and leave the `as` prop unchanged.
+     ```jsx
+     <Avatar badge={{ status: 'busy', render: (_ignored, props) => <MyBadgeComponent {...props} /> }} />
+     ```
+
+### Implementation
+
+The WithSlotRenderFunction type would be expanded to apply the render function support to the `as` prop:
+
+```ts
+type WithSlotRenderFunction<Props> = Props & {
+  as?: (Props extends { as?: string } ? Props['as'] : never) | SlotRenderFunction<Props>;
+
+  // Existing [deprecated] support for a render function on children:
+  children?: (Props extends { children?: unknown } ? Props['children'] : never) | SlotRenderFunction<Props>;
+};
+```
+
+The getSlot function would support `as` being a function:
+
+```ts
+if (typeof asProp === 'function') {
+  const render = asProp as SlotRenderFunction<R[K]>;
+  return [
+    React.Fragment,
+    {
+      children: render(slot, propsWithoutAs),
+    } as unknown as R[K],
+  ];
+}
+```
+
+All existing support for `children` as a render function would still be included, for backwards compatibility.
+
```

Comment:
What will be the adoption strategy for this RFC? I think that it's important to understand it to make a decision.

Reply:
This RFC is a completely new/additive feature: `as` prop can be a render function. So we'd implement the new feature, and update documentation for it. We would keep the existing feature (`children` as a render function) as deprecated, and remove it in the next major release.

Reply:
Added to cons.